### PR TITLE
Feature/14 events

### LINKED ISSUE #14: Write important events

### PR DESCRIPTION
**Pull request by nhinze23**
Resolves #14

### CODE REVIEW

File: controllers/ingressClassCreator.go [hunk 1]
```diff
@@ -35,7 +46,15 @@ func (icc ingressClassCreator) CreateIngressClass(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("failed to check if ingress class [%s] exists: %w", icc.className, err)
 	}
+
+	deployment := &appsv1.Deployment{}
+	err = icc.client.Get(ctx, types.NamespacedName{Name: "k8s-service-discovery-controller-manager", Namespace: icc.namespace}, deployment)
+	if err != nil {
+		return fmt.Errorf("create ingress class: failed to get deployment [%s]: %w", "k8s-service-discovery-controller-manager", err)
```
Comment (Subaro) [inline controllers/ingressClassCreator.go:53]:
```suggestion
		return fmt.Errorf("create ingress class: failed to get deployment [k8s-service-discovery-controller-manager]: %w", err)
```

File: controllers/ingressClassCreator_test.go [hunk 2]
```diff
@@ -12,19 +15,39 @@ import (
...
+		// given
+		clientMock := testclient.NewClientBuilder().WithScheme(getScheme()).WithObjects().Build()
+		creator := NewIngressClassCreator(clientMock, "my-ingress-class", namespace, nil)
+
+		// when
+		err := creator.CreateIngressClass(context.Background())
+
+		// then
+		require.Error(t, err)
+		require.ErrorContains(t, err, "create ingress class: failed to get deployment [k8s-service-discovery-controller-manager]")
```
Comment (Subaro) [inline controllers/ingressClassCreator_test.go:39]: `ErrorContains` also includes and `Error` check. 

```suggestion
		require.ErrorContains(t, err, "create ingress class: failed to get deployment [k8s-service-discovery-controller-manager]")
```

File: controllers/ingressUpdater.go [hunk 3]
```diff
@@ -138,17 +154,24 @@ func (i *ingressUpdater) upsertIngressForCesService(ctx context.Context, cesServ
...
 	}
 
-	return i.upsertDoguIngressObject(ctx, cesService, service)
+	err = i.upsertDoguIngressObject(ctx, cesService, service)
+	if err != nil {
+		return err
+	}
+	i.eventRecorder.Eventf(dogu, corev1.EventTypeNormal, ingressCreationEventReason, "Created regular ingress for service [%s].", cesService.Name)
+
+	return err
```
Comment (Subaro) [inline controllers/ingressUpdater.go:163]: mind the gap :)
```suggestion
	if err != nil {
		return err
	}
	
	i.eventRecorder.Eventf(dogu, corev1.EventTypeNormal, ingressCreationEventReason, "Created regular ingress for service [%s].", cesService.Name)
	return err
```

File: controllers/ingressUpdater.go [hunk 4]
```diff
@@ -138,17 +154,24 @@ func (i *ingressUpdater) upsertIngressForCesService(ctx context.Context, cesServ
...
 	log.FromContext(ctx).Info(fmt.Sprintf("system is in maintenance mode -> create maintenance ingress object for service [%s]", service.GetName()))
 	annotations := map[string]string{ingressRewriteTargetAnnotation: staticContentBackendRewrite}
 
 	err := i.upsertIngressObject(ctx, service, cesService, staticContentBackendName, staticContentBackendPort, annotations)
 	if err != nil {
 		return fmt.Errorf("failed to update ingress object: %w", err)
 	}
+	i.eventRecorder.Eventf(dogu, corev1.EventTypeNormal, ingressCreationEventReason, "Ingress for service [%s] has been updated to maintenance mode.", cesService.Name)
 
 	return nil
```
Comment (Subaro) [inline controllers/ingressUpdater.go:176]: MTG

```suggestion
	if err != nil {
		return fmt.Errorf("failed to update ingress object: %w", err)
	}
	
	i.eventRecorder.Eventf(dogu, corev1.EventTypeNormal, ingressCreationEventReason, "Ingress for service [%s] has been updated to maintenance mode.", cesService.Name)
	return nil
```

File: controllers/ingressUpdater_test.go [hunk 5]
```diff
@@ -139,16 +141,55 @@ func Test_ingressUpdater_UpdateIngressOfService(t *testing.T) {
...
+		creator, creationError := NewIngressUpdater(clientMock, getRegistryMockWithMaintenance(false), myNamespace, myIngressClass, mocks.NewEventRecorder(t))
+		require.NoError(t, creationError)
+
+		// when
+		err := creator.UpsertIngressForService(ctx, &service)
+
+		// then
+		require.Error(t, err)
+		assert.ErrorContains(t, err, "failed to create ingress object for ces service [{Name:test Port:55 Location:/myLocation Pass:/myPass}]")
+		assert.ErrorContains(t, err, "not found")
```
Comment (Subaro) [inline controllers/ingressUpdater_test.go:190]:
```suggestion
		assert.ErrorContains(t, err, "failed to create ingress object for ces service [{Name:test Port:55 Location:/myLocation Pass:/myPass}]")
		assert.ErrorContains(t, err, "not found")
```

File: controllers/serviceController_int_test.go [hunk 6]
```diff
@@ -193,7 +193,7 @@ var _ = Describe("Creating ingress objects with the ingress generator", func() {
 			client, err := kubernetes.NewForConfig(ctrl.GetConfigOrDie())
 			Expect(err).NotTo(HaveOccurred())
 			deploymentWaiter := dogustart.NewDeploymentReadyChecker(client, "my-test-namespace")
-			waitOptions := dogustart.WaitOptions{Timeout: time.Minute * 30, TickRate: time.Millisecond * 200}
+			waitOptions := dogustart.WaitOptions{Timeout: 5, TickRate: 3}
```
Comment (Subaro) [inline controllers/serviceController_int_test.go:196]: Is there a reason for this change? Lowering the timeout should not be a problem. However the tick rate should probably affect the whole time needed to run the unit test by at least 2,8 seconds.

File: controllers/suite_test.go [hunk 7]
```diff
@@ -56,10 +62,15 @@ var _ = BeforeSuite(func() {
...
 
+	testScheme := scheme.Scheme
+	err := doguv1.AddToScheme(testScheme)
+	Expect(err).NotTo(HaveOccurred())
+
 	By("bootstrapping test environment")
 	testEnv = &envtest.Environment{
-		CRDDirectoryPaths:     []string{filepath.Join("..", "config", "bases")},
+		CRDDirectoryPaths:     []string{filepath.Join("..", "vendor", "github.com", "cloudogu", "k8s-dogu-operator", "api", "v1")},
 		ErrorIfCRDPathMissing: false,
```
Comment (Subaro) [inline controllers/suite_test.go:72]: Maybe we should set this to true? `ErrorIfCRDPathMissing`. Sounds reasonable to me.

**Review by Subaro — CHANGES_REQUESTED** [review]

File: controllers/serviceController_int_test.go (hunk 6 above)
Comment (nhinze23) [reply controllers/serviceController_int_test.go]: Reverted tickrate

**Review by Subaro — APPROVED** [review]
LGTM

**Pull request merged**